### ISSUE TITLE
CmdLine/decompiler.exe aborts decompilation with `An internal error occurred while decompiling.`

### ISSUE THREAD

**Issue by mewmew** [uxmal/reko#245]
Analyzed binary: http://ftp.blizzard.com/pub/diablo/patches/pc/drtl109b.zip

The command line driver aborts decompilation with a `An internal error occurred while decompiling.` exception.

```shell
$ decompile.exe drtl109b.exe
...
004038C0: warning: The call or jump table has invalid addresses; stopping.
drtl109b.exe: error: An internal error occurred while decompiling.
    Specified argument was out of the range of valid values.
Parameter name: Linear address 00000000 is not in known segment.
      at Reko.Core.SegmentMap.MapLinearAddressToAddress (UInt64 linearAddress) <0x4082e360 + 0x00197> in <filename unknown>:0 
  at Reko.Analysis.TrashedRegisterFinder.RewriteBlock (Reko.Core.Block block) <0x4082a310 + 0x00383> in <filename unknown>:0 
  at Reko.Analysis.TrashedRegisterFinder.RewriteBasicBlocks () <0x40828ed0 + 0x000ff> in <filename unknown>:0 
  at Reko.Analysis.DataFlowAnalysis.UntangleProcedures () <0x40817a60 + 0x0021b> in <filename unknown>:0 
  at Reko.DecompilerDriver.AnalyzeDataFlow () <0x40816720 + 0x00263> in <filename unknown>:0 
  at Reko.DecompilerDriver.Decompile (System.String filename) <0x4074ced0 + 0x00043> in <filename unknown>:0 
```

Decompiling the same binary using the `WindowsDecompiler.exe` decompilation completes successfully, even after the same warning as been noted in its diagnostics window:

```
004038C0: warning: The call or jump table has invalid addresses; stopping.
```

Discussion ref: https://gitter.im/uxmal/reko?at=5747414210f0fed86f487fac

**Comment by uxmal** [comment]
I've confirmed this on my end. Next step is to create a unit test that reproduces the issue.

**Comment by uxmal** [comment]
I've found the cause of the bug. Fixing it revealed another hiding behind it. Stay tuned....

**Comment by mewmew** [comment]
> I've found the cause of the bug. Fixing it revealed another hiding behind it. Stay tuned....

Hopefully it won't be turtles all the way down :)

**Comment by uxmal** [comment]
OK, this fixes the immediate problem you've reported. The other is a deeper issue which has to do with the eager way Reko builds the CFGraphs for procedures. That fix requires some major reworking, and I won't be doing that at 2am :) 

Please verify and close.

**Comment by mewmew** [comment]
I can verify that the original issue seems to have been resolve, and as you mentioned, another one showed up in its place.

For the same input file using `decompiler.exe`, I now get to the point where the following errors are displayed and the `decompiler.exe` process seems to go into an infinite loop, hogging the CPU indefinitely. I haven't verified letting it run more than 1 minute or so, but from the looks of it, no progress is made.

```
l0040ABFD: error: An error occurred while processing instruction at address 0040AC19.
    Expected stack depth to be negative.
      at Reko.Analysis.ExpressionPropagator.GetStackDepthBeforeCall () <0x40c635a0 + 0x001db> in <filename unknown>:0 
  at Reko.Analysis.ExpressionPropagator.VisitCallInstruction (Reko.Core.Code.CallInstruction ci) <0x40c633d0 + 0x0003b> in <filename unknown>:0 
  at Reko.Core.Code.CallInstruction.Accept[T] (InstructionVisitor`1 visitor) <0x40c5ae10 + 0x00030> in <filename unknown>:0 
  at Reko.Analysis.TrashedRegisterFinder.RewriteBlock (Reko.Core.Block block) <0x40c610b0 + 0x00205> in <filename unknown>:0 
l0040ABFD: error: An error occurred while processing instruction at address 0040AC26.
    Expected stack depth to be negative.
      at Reko.Analysis.ExpressionPropagator.GetStackDepthBeforeCall () <0x40c635a0 + 0x001db> in <filename unknown>:0 
  at Reko.Analysis.ExpressionPropagator.VisitCallInstruction (Reko.Core.Code.CallInstruction ci) <0x40c633d0 + 0x0003b> in <filename unknown>:0 
  at Reko.Core.Code.CallInstruction.Accept[T] (InstructionVisitor`1 visitor) <0x40c5ae10 + 0x00030> in <filename unknown>:0 
  at Reko.Analysis.TrashedRegisterFinder.RewriteBlock (Reko.Core.Block block) <0x40c610b0 + 0x00205> in <filename unknown>:0 
l0040ABFD: error: An error occurred while processing instruction at address 0040AC2E.
    Expected stack depth to be negative.
      at Reko.Analysis.ExpressionPropagator.GetStackDepthBeforeCall () <0x40c635a0 + 0x001db> in <filename unknown>:0 
  at Reko.Analysis.ExpressionPropagator.VisitCallInstruction (Reko.Core.Code.CallInstruction ci) <0x40c633d0 + 0x0003b> in <filename unknown>:0 
  at Reko.Core.Code.CallInstruction.Accept[T] (InstructionVisitor`1 visitor) <0x40c5ae10 + 0x00030> in <filename unknown>:0 
  at Reko.Analysis.TrashedRegisterFinder.RewriteBlock (Reko.Core.Block block) <0x40c610b0 + 0x00205> in <filename unknown>:0 
l00433883: error: An error occurred while processing instruction at address 00433888.
    The given key was not present in the dictionary.
      at System.ThrowHelper.ThrowKeyNotFoundException () <0x7f04c0494b60 + 0x00015> in <filename unknown>:0 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].get_Item (System.Collections.Generic.TKey key) <0x7f04c030cdb0 + 0x0004a> in <filename unknown>:0 
  at Reko.Analysis.ExpressionPropagator.GetStackDepthBeforeCall () <0x40c635a0 + 0x00064> in <filename unknown>:0 
  at Reko.Analysis.ExpressionPropagator.VisitCallInstruction (Reko.Core.Code.CallInstruction ci) <0x40c633d0 + 0x0003b> in <filename unknown>:0 
  at Reko.Core.Code.CallInstruction.Accept[T] (InstructionVisitor`1 visitor) <0x40c5ae10 + 0x00030> in <filename unknown>:0 
  at Reko.Analysis.TrashedRegisterFinder.RewriteBlock (Reko.Core.Block block) <0x40c610b0 + 0x00205> in <filename unknown>:0 
l0046B794_thunk_fn0046E3DE: error: An error occurred while processing instruction at address 0046B794.
    The given key was not present in the dictionary.
      at System.ThrowHelper.ThrowKeyNotFoundException () <0x7f04c0494b60 + 0x00015> in <filename unknown>:0 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].get_Item (System.Collections.Generic.TKey key) <0x7f04c030cdb0 + 0x0004a> in <filename unknown>:0 
  at Reko.Analysis.ExpressionPropagator.GetStackDepthBeforeCall () <0x40c635a0 + 0x00064> in <filename unknown>:0 
  at Reko.Analysis.ExpressionPropagator.VisitCallInstruction (Reko.Core.Code.CallInstruction ci) <0x40c633d0 + 0x0003b> in <filename unknown>:0 
  at Reko.Core.Code.CallInstruction.Accept[T] (InstructionVisitor`1 visitor) <0x40c5ae10 + 0x00030> in <filename unknown>:0 
  at Reko.Analysis.TrashedRegisterFinder.RewriteBlock (Reko.Core.Block block) <0x40c610b0 + 0x00205> in <filename unknown>:0 
```

I'll close this issue, as the issue it was intended for has been resolved. Thanks for taking your time to look at it.

Have you opened up a new issue for the CFG related one, or are you simply keeping a mental note for now?

**Comment by uxmal** [comment]
This issue is well known. The fix requires some invasive rearchitecting of the way Reko performs scanning of images. This work is being done in the scanner-development branch. For now, a workaround for this kind of issue is marking addresses which are jumped to by instructions in two different procedures as a user procedure. Reko will be able to cope with itself once scanner-development is complete.

**Comment by mewmew** [comment]
> This issue is well known. The fix requires some invasive rearchitecting of the way Reko performs scanning of images. This work is being done in the scanner-development branch.

Lovely to hear. I'm especially glad that you focus on fix core issues rather than symptoms. Having been in the game for 10 year by now, I'm sure you've long since figured out this is the only way to keep a clean architecture and consistent design, instead of bolting on patch after patch. I am looking forward to meeting you in person later this summer. Would love to continue discussing these topics in person.

But for now, it's time for bed :) Good night John
